include/swift/SIL/SILNode.h

```diff
@@ -113,15 +113,20 @@ class alignas(8) SILNode {
 
 protected:
   SILNode(SILNodeKind kind, SILNodeStorageLocation storageLoc)
-    : Kind(unsigned(kind)),
-      StorageLoc(unsigned(storageLoc)),
-      IsCanonical(storageLoc == SILNodeStorageLocation::Instruction ||
-                  !hasMultipleSILNodes(kind)) {}
+      : Kind(unsigned(kind)), StorageLoc(unsigned(storageLoc)),
+        IsCanonical(storageLoc == SILNodeStorageLocation::Instruction ||
+                    // TODO: Ask John about why he is using
+                    // hasMultipleSILNodeBaseClasses. Seems weird.
+                    (!hasMultipleSILNodeBaseClasses(kind) &&
+                     kind != SILNodeKind::MultipleValueInstructionResult)) {}
```

Comment:
MultipleValueInstructions still contain multiple SILNodes (unless they have zero results) and their multi-result nodes are still not canonical.  The existing function name and logic was correct here.  hasMultipleSILNodes does not need to be updated for MultipleValueInstructions because the SILNodeKind stored in a multi-result node is not the same as its instruction's SILNodeKind.

Reply:
Let me elaborate:

What is hasMultipleSILNodes actually supposed to do? If it is supposed to return true if a node contains multiple SIL nodes, then I would assume that it would return true for MultipleValueInstruction and false for all other SILNodes. In contrast, hasMultipleSILNodes() returns true for SingleValueInstruction and false for all other SILNodes. How does the name of the method correspond to its function. The only thing I could think of is that it is an instruction with multiple SILNode base classes (i.e. inheritance diamonds).

The reason why I added the special case for MultipleValueInstructionResult is as mentioned above hasMultipleSILNodes returns false for it, so MultipleValueInstructionResult are considered canonical. =><=.

Reply:
@rjmccall (Keep in mind, I basically just put something in here that worked with the intention of asking you during the review).

Reply:
Well, there are two separate purposes.

The first is that some code may rely on getting a unique SILNode for an instruction; such code should always be using the SILNode from the SILInstruction rather than the SILNode associated with any values produced by the instruction.  This is what getCanonicalSILNode() cares about.

The second is that we use a single SILNodeKind value for both the value node and the instruction node of a SingleValueInstruction, so you can't safely static_cast a SILNode* to SILInstruction*, ValueBase*, or any other subclass just because the SILNodeKind says it's a kind of SingleValueInstruction.  You're correct that this is related to the diamond inheritance.

Reply:
Lets talk about this in person since you are in town?

TBH I find that our use of multiple inheritence is brittle. I wonder if we could use composition + tricks (i.e. conversions or changes to dyn_cast ala being able to dyn_cast SILValue) to simplify things. Maybe introduce a SingleValueResult that is produced by SingleValueInstruction. Then we have just a much simpler model from a cpp perspective.